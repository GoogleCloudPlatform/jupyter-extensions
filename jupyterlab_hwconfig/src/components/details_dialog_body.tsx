/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { stylesheet, classes } from 'typestyle';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableRow from '@material-ui/core/TableRow';
import Paper from '@material-ui/core/Paper';
import { STYLES } from '../data/styles';
import { MAPPED_ATTRIBUTES, Details } from '../data/data';
import { ActionBar } from './action_bar';

const useStyles = makeStyles({
  table: {
    minWidth: 350,
  },
  container: {
    boxShadow: 'none',
    elevation: 0,
    backgroundColor: 'var(--jp-layout-color1)',
  },
  textColor: {
    color: 'var(--jp-ui-font-color1)',
  },
});

function SimpleTable(details: Details) {
  const tableClasses = useStyles();

  return (
    <TableContainer className={tableClasses.container} component={Paper}>
      <Table className={tableClasses.table} aria-label="simple table">
        <TableBody>
          {MAPPED_ATTRIBUTES.map(am => (
            <TableRow key={am.label}>
              <TableCell
                className={tableClasses.textColor}
                component="th"
                scope="row"
              >
                {am.label}
              </TableCell>
              <TableCell className={tableClasses.textColor} align="right">
                {am.mapper(details)}
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

interface Props {
  details: Details;
  receivedError: boolean;
  onDialogClose: () => void;
  onUpdate: () => void;
}

const DIALOG_STYLES = stylesheet({
  headingPadding: {
    paddingBottom: '15px',
  },
});

function isLoadingDetails(details: boolean, receivedError: boolean): boolean {
  return !(details || receivedError);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export function DetailsDialogBody(props: Props) {
  const { details, receivedError, onDialogClose, onUpdate } = props;
  return (
    <dl className={STYLES.containerPadding}>
      <p className={classes(STYLES.heading, DIALOG_STYLES.headingPadding)}>
        Notebook VM Details
      </p>
      {isLoadingDetails(Boolean(details), receivedError) ? (
        <div className={STYLES.containerSize}>
          <p className={STYLES.paragraph}>Retrieving GCE VM details...</p>
        </div>
      ) : receivedError ? (
        <div className={STYLES.containerSize}>
          <p className={STYLES.paragraph}>
            Unable to retrieve GCE VM details, please check your server logs
          </p>
        </div>
      ) : (
        SimpleTable(details)
      )}
      <ActionBar
        primaryLabel="Update"
        secondaryLabel="Close"
        onPrimaryClick={onUpdate}
        onSecondaryClick={onDialogClose}
        primaryDisabled={
          receivedError || isLoadingDetails(Boolean(details), receivedError)
        }
      />
    </dl>
  );
}
