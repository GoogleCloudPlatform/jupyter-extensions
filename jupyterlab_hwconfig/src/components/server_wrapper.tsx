/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ServerConnection } from '@jupyterlab/services';
import { Details } from '../data/data';

export class ServerWrapper {
  private readonly serverSettings: ServerConnection.ISettings;
  private readonly url: string;
  constructor(
    serverSettings: ServerConnection.ISettings = ServerConnection.defaultSettings
  ) {
    this.serverSettings = serverSettings;
    this.url = this.serverSettings.baseUrl + `gcp/v1/details`;
  }

  async getUtilizationData(): Promise<Details> {
    const response = await ServerConnection.makeRequest(
      this.url,
      {},
      this.serverSettings
    );
    if (!response.ok) {
      throw new Error();
    }
    const data = await response.json();
    return { ...data };
  }
}
