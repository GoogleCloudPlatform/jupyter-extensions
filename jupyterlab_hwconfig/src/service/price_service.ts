/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ServerConnection } from '@jupyterlab/services';
import {
  HardwareConfiguration,
  REGIONS,
  HOURS_PER_MONTH,
  NO_SUSTAINED_DISCOUNT_PREFIXES,
} from '../data/data';

/** The monthly price catalog for region and sku */
type Catalog = Map<string, Map<string, number>>;

export class PriceService {
  private catalog: Catalog;

  // Exposed for testing purposes only
  set _catalog(catalog: Catalog) {
    this.catalog = catalog;
  }

  private async buildCatalog(): Promise<void> {
    // Only attempt to build the catalog once
    if (this.catalog) {
      return;
    }
    try {
      this.catalog = new Map();
      const response = await ServerConnection.makeRequest(
        `${ServerConnection.defaultSettings.baseUrl}gcp/v1/pricelist`,
        {},
        ServerConnection.defaultSettings
      );
      if (!response.ok) {
        throw { result: await response.text() };
      }
      const fullCatalog = await response.json();
      const pricelist = fullCatalog.gcp_price_list;
      Object.keys(pricelist).forEach(sku => {
        if (sku.endsWith('PREEMPTIBLE')) return;
        if (
          sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-N1') ||
          sku.startsWith('CP-COMPUTEENGINE-VMIMAGE-E2')
        ) {
          this.populateCatalog(
            sku.substring(25).toLowerCase(),
            pricelist[sku],
            true
          );
        } else if (sku.startsWith('GPU_')) {
          this.populateCatalog(sku.substring(4), pricelist[sku], true);
        } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-CAPACITY') {
          this.populateCatalog('pd-standard', pricelist[sku], false);
        } else if (sku === 'CP-COMPUTEENGINE-STORAGE-PD-SSD') {
          this.populateCatalog('pd-ssd', pricelist[sku], false);
        }
      });
    } catch (err) {
      console.error(`Unable to retrieve pricelist.`);
    }
  }

  // tslint:disable-next-line:no-any The json object returned is not typed.
  private populateCatalog(sku: string, prices: any, timeBased: boolean) {
    Object.keys(prices).forEach(region => {
      const regionCatalog = this.getRegionCatalog(region);
      if (regionCatalog) {
        regionCatalog.set(
          sku,
          prices[region] * (timeBased ? HOURS_PER_MONTH : 1.0)
        );
      }
    });
  }

  private getRegionCatalog(region: string): Map<string, number> | undefined {
    if (REGIONS.find(r => r.value === region)) {
      if (!this.catalog.has(region)) {
        this.catalog.set(region, new Map<string, number>());
      }
      return this.catalog.get(region);
    }
    return undefined;
  }

  private getRegionFromZone(zone: string): string {
    return zone.split('-', 2).join('-');
  }

  /**
   * Returns estimated total monthly price for a given instance to be created in a
   * certain region. If catalog doesn't contain relevant resource information
   * returns undefined.
   */
  async getPrice(
    zone: string,
    config: HardwareConfiguration
  ): Promise<number | undefined> {
    await this.buildCatalog();

    const regionCatalog = this.getRegionCatalog(this.getRegionFromZone(zone!));
    if (!regionCatalog) return undefined;

    let total = 0;
    let hasSustainedUseDiscount = true;

    if (config.machineType) {
      const machineTypeValue = config.machineType.name;
      total += regionCatalog.get(machineTypeValue);

      hasSustainedUseDiscount = !NO_SUSTAINED_DISCOUNT_PREFIXES.some(prefix =>
        machineTypeValue.startsWith(prefix)
      );
    }

    if (config.attachGpu && config.gpuType && config.gpuCount) {
      total += regionCatalog.get(config.gpuType) * parseInt(config.gpuCount);
    }

    if (config.diskType && config.diskSizeGb) {
      total += regionCatalog.get(config.diskType) * parseInt(config.diskSizeGb);
    }

    if (hasSustainedUseDiscount) {
      total *= 0.7;
    }

    return isNaN(total) ? undefined : total;
  }
}
