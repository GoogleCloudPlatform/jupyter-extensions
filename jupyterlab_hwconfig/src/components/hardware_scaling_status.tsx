/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as csstips from 'csstips';
import * as React from 'react';
import { stylesheet, classes } from 'typestyle';
import LinearProgress from '@material-ui/core/LinearProgress';
import { createStyles, withStyles, Theme } from '@material-ui/core/styles';
import { STYLES } from '../data/styles';
import { HardwareConfiguration } from '../data/data';
import { MachineTypeConfiguration } from '../data/machine_types';
import { ActionBar } from './action_bar';
import { NotebooksService, Instance } from '../service/notebooks_service';
import { ServerWrapper } from './server_wrapper';
import { ErrorPage } from './error_page';
import { displayInstance } from './instance_details_message';

const BorderLinearProgress = withStyles((theme: Theme) =>
  createStyles({
    root: {
      height: 10,
      width: 450,
      borderRadius: 15,
      margin: '10px 0px',
    },
    colorPrimary: {
      backgroundColor:
        theme.palette.grey[theme.palette.type === 'light' ? 200 : 700],
    },
    bar: {
      backgroundColor: '#1a90ff',
      borderRadius: 5,
    },
  })
)(LinearProgress);

const STATUS_STYLES = stylesheet({
  container: {
    ...csstips.vertical,
    alignItems: 'center',
    backgroundColor: 'var(--jp-layout-color1)',
  },
  containerPadding: {
    padding: '50px 75px 15px 75px',
  },
  containerBottomPadding: {
    paddingBottom: '20px',
  },
  image: {
    marginBottom: '40px',
    height: 240,
    width: 300,
  },
  bottomText: {
    marginTop: '50px',
    fontSize: '12px',
    color: 'var(--jp-ui-font-color2)',
  },
});

export enum Status {
  'Authorizing' = 0,
  'Stopping notebook instance' = 1,
  'Updating machine configuration' = 2,
  'Updating GPU configuration' = 3,
  'Restarting notebook instance' = 4,
  'Refreshing session' = 5,
  'Complete' = 6,
  'Error' = 7,
}

const statusInfo = [
  'Please complete the OAuth 2.0 authorization steps in the popup',
  'Shutting down your notebook instance for reshaping.',
  'Reshaping machine configuration to match your selection.',
  'Reshaping GPU configuration to match your selection.',
  'Restarting your newly configured notebook instance.',
  'Refreshing your JupyterLab session to reflect your new configuration.',
  'Operation complete. Enjoy your newly configured instance!',
  'An error has occured, please try again later.',
];

export enum ErrorType {
  STOP = 'Stop',
  RESHAPING = 'Reshape',
  START = 'Start',
  REFRESH = 'Refresh',
}

export interface ConfigurationError {
  errorType: ErrorType;
  errorValue: any;
}

interface Props {
  hardwareConfiguration: HardwareConfiguration;
  notebookService: NotebooksService;
  onDialogClose: () => void;
  onCompletion: () => void;
  detailsServer: ServerWrapper;
  authTokenRetrieval: () => Promise<string>;
  machineTypes: MachineTypeConfiguration[];
}

interface State {
  status: Status;
  instanceDetails: Instance;
  error: ConfigurationError;
}

export class HardwareScalingStatus extends React.Component<Props, State> {
  private readonly NUM_RETRIES = 20;

  constructor(props: Props) {
    super(props);
    this.state = {
      status: Status.Authorizing,
      instanceDetails: null,
      error: null,
    };
  }

  private preventPageClose(event) {
    event.preventDefault();
    event.returnValue = '';
  }

  private updateStatus(status: Status, instanceDetails: Instance) {
    this.setState({
      status,
      instanceDetails,
    });
  }

  private showError(
    errorType: ErrorType,
    errorValue: any,
    newInstanceDetails?: Instance
  ) {
    this.setState({
      status: Status.Error,
      instanceDetails: newInstanceDetails
        ? newInstanceDetails
        : this.state.instanceDetails,
      error: {
        errorType,
        errorValue,
      },
    });
  }

  private updateError(
    errorType: ErrorType,
    errorValue: any,
    newInstanceDetails?: Instance
  ) {
    this.setState({
      instanceDetails: newInstanceDetails
        ? newInstanceDetails
        : this.state.instanceDetails,
      error: {
        errorType,
        errorValue,
      },
    });
  }

  private async waitForServer() {
    for (let tries = 0; tries < this.NUM_RETRIES; tries++) {
      try {
        await this.props.detailsServer.getUtilizationData();
        break;
      } catch (err) {
        if (tries === this.NUM_RETRIES - 1) {
          this.showError(ErrorType.REFRESH, err);
        }
        await new Promise(resolve => setTimeout(resolve, 5000));
      }
    }
  }

  async componentDidMount() {
    const { notebookService, authTokenRetrieval } = this.props;
    try {
      const token = await authTokenRetrieval();
      this.setState({
        status: Status['Stopping notebook instance'],
      });
      notebookService.setAuthToken(token);
    } catch (err) {
      this.setState({ status: Status.Error });
      console.log(err);
    }
    window.addEventListener('beforeunload', this.preventPageClose);
  }

  async componentDidUpdate(prevProps, prevState) {
    const { status, error, instanceDetails } = this.state;
    const { hardwareConfiguration, notebookService, onCompletion } = this.props;
    const { machineType, attachGpu, gpuType, gpuCount } = hardwareConfiguration;

    if (prevState.status !== status) {
      switch (status) {
        case Status['Stopping notebook instance']: {
          try {
            const stopResult = await notebookService.stop();
            this.updateStatus(
              Status['Updating machine configuration'],
              stopResult
            );
          } catch (err) {
            this.showError(ErrorType.STOP, err);
          }
          break;
        }

        case Status['Updating machine configuration']: {
          let setMachineTypeResult = instanceDetails;
          try {
            setMachineTypeResult = await notebookService.setMachineType(
              machineType.name
            );
          } catch (err) {
            this.updateError(ErrorType.RESHAPING, err);
          }
          const nextStatus = attachGpu
            ? 'Updating GPU configuration'
            : 'Restarting notebook instance';
          this.updateStatus(Status[nextStatus], setMachineTypeResult);
          break;
        }

        case Status['Updating GPU configuration']: {
          let setGpuTypeResult = instanceDetails;
          try {
            setGpuTypeResult = await notebookService.setAccelerator(
              gpuType,
              gpuCount
            );
          } catch (err) {
            this.updateError(ErrorType.RESHAPING, err);
          }
          this.updateStatus(
            Status['Restarting notebook instance'],
            setGpuTypeResult
          );
          break;
        }

        case Status['Restarting notebook instance']: {
          try {
            const startResult = await notebookService.start();
            this.updateStatus(Status['Refreshing session'], startResult);
          } catch (err) {
            this.showError(ErrorType.START, err);
          }
          break;
        }

        case Status['Refreshing session']: {
          await this.waitForServer();
          onCompletion();
          this.setState({
            status:
              error && error.errorType === ErrorType.RESHAPING
                ? Status.Error
                : Status.Complete,
          });
          break;
        }

        default:
          break;
      }
    }
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.preventPageClose);
  }

  render() {
    const { status, error, instanceDetails } = this.state;
    const progressValue = ((status + 1) / 6) * 100;
    const { onDialogClose, machineTypes } = this.props;
    return status === Status['Error'] ? (
      <ErrorPage
        onDialogClose={onDialogClose}
        error={error}
        instanceDetails={instanceDetails}
        machineTypes={machineTypes}
      />
    ) : (
      <div
        className={classes(
          STATUS_STYLES.containerPadding,
          STATUS_STYLES.container
        )}
      >
        <span className={classes(STATUS_STYLES.image, 'jp-CaipIcon')} />
        <p className={STYLES.heading}>
          {status < 6 ? `${Status[status]} (${status + 1}/6)` : Status[status]}
        </p>
        {status === Status['Complete'] ? (
          <div
            className={
              (STATUS_STYLES.container, STATUS_STYLES.containerBottomPadding)
            }
          >
            <p className={STYLES.paragraph}>{statusInfo[status]}</p>
            {status === Status['Complete'] &&
              displayInstance(
                instanceDetails,
                machineTypes,
                'Your new configuration:'
              )}
            <ActionBar onPrimaryClick={onDialogClose} primaryLabel="Close" />
          </div>
        ) : (
          <div className={STATUS_STYLES.container}>
            <BorderLinearProgress variant="determinate" value={progressValue} />
            <p className={STYLES.paragraph}>{statusInfo[status]}</p>
            <p className={STATUS_STYLES.bottomText}>
              Don't close the browser tab before the update process is finished
            </p>
          </div>
        )}
      </div>
    );
  }
}
