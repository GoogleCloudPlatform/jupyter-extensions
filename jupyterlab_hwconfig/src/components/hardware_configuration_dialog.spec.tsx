/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { shallow } from 'enzyme';
import { ConfirmationPage } from './confirmation_page';
import { DetailsDialogBody } from './details_dialog_body';
import {
  HardwareConfigurationDialog,
  View,
} from './hardware_configuration_dialog';
import { HardwareScalingForm } from './hardware_scaling_form';
import { HardwareScalingStatus } from './hardware_scaling_status';
import { ServerWrapper } from './server_wrapper';
import {
  NO_ACCELERATOR_TYPE,
  NO_ACCELERATOR_COUNT,
} from '../data/accelerator_types';
import { NotebooksService } from '../service/notebooks_service';
import { PriceService } from '../service/price_service';
import { DETAILS } from '../test_helpers';

const CONFIGURATION = {
  machineType: {
    description: '4 vCPU, 15 GB RAM',
    name: 'n1-standard-4',
  },
  attachGpu: false,
  gpuType: NO_ACCELERATOR_TYPE,
  gpuCount: NO_ACCELERATOR_COUNT,
};

describe('HardwareScalingForm', () => {
  const mockOnClose = jest.fn();
  const mockOnCompletion = jest.fn();
  const mockDetailsServer = {} as ServerWrapper;
  const mockNotebookService = {} as NotebooksService;
  const mockPriceService = {} as PriceService;

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('Opens on details view', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );

    expect(hardwareConfigurationDialog.state('view')).toEqual(View.DETAILS);
    expect(hardwareConfigurationDialog).toMatchSnapshot('Details View');
  });

  it('Shows form view after details view', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );

    hardwareConfigurationDialog.find(DetailsDialogBody).prop('onUpdate')();
    expect(hardwareConfigurationDialog.state('view')).toEqual(View.FORM);
    expect(hardwareConfigurationDialog).toMatchSnapshot('Form View');
  });

  it('Shows confirmation view after form view', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );

    hardwareConfigurationDialog.setState({ view: View.FORM });
    hardwareConfigurationDialog.find(HardwareScalingForm).prop('onSubmit')(
      CONFIGURATION
    );
    expect(hardwareConfigurationDialog.state('view')).toEqual(
      View.CONFIRMATION
    );
    expect(hardwareConfigurationDialog.state('hardwareConfiguration')).toEqual(
      CONFIGURATION
    );
    expect(hardwareConfigurationDialog).toMatchSnapshot('Confirmation View');
  });

  it('Shows status view after confirmation view', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );

    hardwareConfigurationDialog.setState({
      view: View.CONFIRMATION,
      hardwareConfiguration: CONFIGURATION,
    });
    hardwareConfigurationDialog.find(ConfirmationPage).prop('onSubmit')();
    expect(hardwareConfigurationDialog.state('view')).toEqual(View.STATUS);
    expect(hardwareConfigurationDialog).toMatchSnapshot('Status View');
  });

  it('Closes dialog from details view on close', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );

    hardwareConfigurationDialog.find(DetailsDialogBody).prop('onDialogClose')();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('Closes dialog from form view on close', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );
    hardwareConfigurationDialog.setState({
      view: View.FORM,
      hardwareConfiguration: CONFIGURATION,
    });

    hardwareConfigurationDialog
      .find(HardwareScalingForm)
      .prop('onDialogClose')();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('Closes dialog from confirmation view on close', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );
    hardwareConfigurationDialog.setState({
      view: View.CONFIRMATION,
      hardwareConfiguration: CONFIGURATION,
    });

    hardwareConfigurationDialog.find(ConfirmationPage).prop('onDialogClose')();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('Closes dialog from status view on close', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );
    hardwareConfigurationDialog.setState({
      view: View.STATUS,
      hardwareConfiguration: CONFIGURATION,
    });

    hardwareConfigurationDialog
      .find(HardwareScalingStatus)
      .prop('onDialogClose')();
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('Calls onCompletion function when hardware scaling status completes', async () => {
    const hardwareConfigurationDialog = shallow(
      <HardwareConfigurationDialog
        open={true}
        receivedError={false}
        details={DETAILS}
        detailsServer={mockDetailsServer}
        notebookService={mockNotebookService}
        priceService={mockPriceService}
        onClose={mockOnClose}
        onCompletion={mockOnCompletion}
      />
    );
    hardwareConfigurationDialog.setState({
      view: View.STATUS,
      hardwareConfiguration: CONFIGURATION,
    });

    hardwareConfigurationDialog
      .find(HardwareScalingStatus)
      .prop('onCompletion')();
    expect(mockOnCompletion).toHaveBeenCalledTimes(1);
  });
});
