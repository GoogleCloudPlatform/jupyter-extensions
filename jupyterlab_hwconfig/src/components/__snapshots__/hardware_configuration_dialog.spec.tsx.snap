// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HardwareScalingForm Opens on details view: Details View 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <DetailsDialogBody
    details={
      Object {
        "acceleratorTypes": Array [
          Object {
            "description": "Nvidia Tesla K80",
            "maximumCardsPerInstance": 4,
            "name": "NVIDIA_TESLA_K80",
          },
          Object {
            "description": "Nvidia Tesla T4",
            "maximumCardsPerInstance": 8,
            "name": "NVIDIA_TESLA_T4",
          },
        ],
        "gpu": Object {
          "count": "0",
          "cuda_version": "10.1",
          "driver_version": "418.87.01",
          "gpu": 0,
          "memory": 0,
          "name": null,
          "temperature": 0,
        },
        "instance": Object {
          "attributes": Object {
            "framework": "PyTorch:1.4",
            "title": "PyTorch/fastai/CUDA10.0",
            "version": "44",
          },
          "cpuPlatform": "Intel Broadwell",
          "id": 127546640929027970,
          "image": "projects/deeplearning-platform-release/global/images/pytorch-1-4-cu101-notebooks-20200302",
          "machineType": Object {
            "description": "4 vCPU, 15 GB RAM",
            "name": "n1-standard-4",
          },
          "name": "pytorch",
          "zone": "projects/123456/zones/us-west1-b",
        },
        "machineTypes": Array [
          Object {
            "base": Object {
              "text": "N1 Standard",
              "value": "n1-",
            },
            "configurations": Array [
              Object {
                "text": "2 vCPU, 7.5 GB RAM",
                "value": "n1-standard-2",
              },
              Object {
                "text": "4 vCPU, 15 GB RAM",
                "value": "n1-standard-4",
              },
            ],
          },
          Object {
            "base": Object {
              "text": "N2 Standard",
              "value": "n2-",
            },
            "configurations": Array [
              Object {
                "text": "4 vCPU, 15 GB RAM",
                "value": "n2-standard-4",
              },
            ],
          },
        ],
        "project": Object {
          "numericProjectId": 123456,
          "projectId": "test-project",
        },
        "utilization": Object {
          "cpu": 50,
          "memory": 16,
        },
      }
    }
    onDialogClose={[MockFunction]}
    onUpdate={[Function]}
    receivedError={false}
  />
</WithStyles(ForwardRef(Dialog))>
`;

exports[`HardwareScalingForm Shows confirmation view after form view: Confirmation View 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <ConfirmationPage
    currentConfiguration={
      Object {
        "attachGpu": false,
        "gpuCount": "0",
        "gpuType": "ACCELERATOR_TYPE_UNSPECIFIED",
        "machineType": Object {
          "description": "4 vCPU, 15 GB RAM",
          "name": "n1-standard-4",
        },
      }
    }
    formData={
      Object {
        "attachGpu": false,
        "gpuCount": "0",
        "gpuType": "ACCELERATOR_TYPE_UNSPECIFIED",
        "machineType": Object {
          "description": "4 vCPU, 15 GB RAM",
          "name": "n1-standard-4",
        },
      }
    }
    onDialogClose={[MockFunction]}
    onSubmit={[Function]}
  />
</WithStyles(ForwardRef(Dialog))>
`;

exports[`HardwareScalingForm Shows form view after details view: Form View 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <HardwareScalingForm
    details={
      Object {
        "acceleratorTypes": Array [
          Object {
            "description": "Nvidia Tesla K80",
            "maximumCardsPerInstance": 4,
            "name": "NVIDIA_TESLA_K80",
          },
          Object {
            "description": "Nvidia Tesla T4",
            "maximumCardsPerInstance": 8,
            "name": "NVIDIA_TESLA_T4",
          },
        ],
        "gpu": Object {
          "count": "0",
          "cuda_version": "10.1",
          "driver_version": "418.87.01",
          "gpu": 0,
          "memory": 0,
          "name": null,
          "temperature": 0,
        },
        "instance": Object {
          "attributes": Object {
            "framework": "PyTorch:1.4",
            "title": "PyTorch/fastai/CUDA10.0",
            "version": "44",
          },
          "cpuPlatform": "Intel Broadwell",
          "id": 127546640929027970,
          "image": "projects/deeplearning-platform-release/global/images/pytorch-1-4-cu101-notebooks-20200302",
          "machineType": Object {
            "description": "4 vCPU, 15 GB RAM",
            "name": "n1-standard-4",
          },
          "name": "pytorch",
          "zone": "projects/123456/zones/us-west1-b",
        },
        "machineTypes": Array [
          Object {
            "base": Object {
              "text": "N1 Standard",
              "value": "n1-",
            },
            "configurations": Array [
              Object {
                "text": "2 vCPU, 7.5 GB RAM",
                "value": "n1-standard-2",
              },
              Object {
                "text": "4 vCPU, 15 GB RAM",
                "value": "n1-standard-4",
              },
            ],
          },
          Object {
            "base": Object {
              "text": "N2 Standard",
              "value": "n2-",
            },
            "configurations": Array [
              Object {
                "text": "4 vCPU, 15 GB RAM",
                "value": "n2-standard-4",
              },
            ],
          },
        ],
        "project": Object {
          "numericProjectId": 123456,
          "projectId": "test-project",
        },
        "utilization": Object {
          "cpu": 50,
          "memory": 16,
        },
      }
    }
    onDialogClose={[MockFunction]}
    onSubmit={[Function]}
    priceService={Object {}}
  />
</WithStyles(ForwardRef(Dialog))>
`;

exports[`HardwareScalingForm Shows status view after confirmation view: Status View 1`] = `
<WithStyles(ForwardRef(Dialog))
  open={true}
>
  <HardwareScalingStatus
    authTokenRetrieval={[Function]}
    detailsServer={Object {}}
    hardwareConfiguration={
      Object {
        "attachGpu": false,
        "gpuCount": "0",
        "gpuType": "ACCELERATOR_TYPE_UNSPECIFIED",
        "machineType": Object {
          "description": "4 vCPU, 15 GB RAM",
          "name": "n1-standard-4",
        },
      }
    }
    machineTypes={
      Array [
        Object {
          "base": Object {
            "text": "N1 Standard",
            "value": "n1-",
          },
          "configurations": Array [
            Object {
              "text": "2 vCPU, 7.5 GB RAM",
              "value": "n1-standard-2",
            },
            Object {
              "text": "4 vCPU, 15 GB RAM",
              "value": "n1-standard-4",
            },
          ],
        },
        Object {
          "base": Object {
            "text": "N2 Standard",
            "value": "n2-",
          },
          "configurations": Array [
            Object {
              "text": "4 vCPU, 15 GB RAM",
              "value": "n2-standard-4",
            },
          ],
        },
      ]
    }
    notebookService={Object {}}
    onCompletion={[MockFunction]}
    onDialogClose={[MockFunction]}
  />
</WithStyles(ForwardRef(Dialog))>
`;
